### PR TITLE
interfaces/docker: use implicitOnClassic: true

### PR DESCRIPTION
**Pull request by jdstrand**
The docker interface currently can only be used with app snaps. This
works fine when the docker snap is installed, but doesn't work when a
snap plugs docker and tries to use docker from another package format.

Add implicitOnClassic: true to the setup. This is a breaking change for
the docker snap itself since to provides the docker slot and now so will
the system, thus causing an ambiguity for snapd which causes the
docker-cli plug to not auto-connect. With this change, the docker snap
needs to be updated to have (or similar):
```
  plugs:
    docker:
      allow-auto-connection:
        slot-snap-id:
          - <snap id of the docker snap>
```
The store will have to be queried to for any other snaps that have
auto-connections and their snap declarations will need to be updated to
use greedy plugs (or similar):
```
  plugs:
    docker:
      allow-auto-connection:
        slot-snap-type:
          - app
          - core
        slots-per-plug: *
```
(slot-snap-type is needed due to LP: #1873282). In this manner, snaps
will be able to auto-connect regardless of if the docker snap is
installed or not.

References:
https://forum.snapcraft.io/t/request-for-classic-confinement-for-package-wilfred/14912/49

### CODE REVIEW

**Comment by jdstrand** [comment]
I've submitted this, but marked it as blocked since a) I'd like @pedronis to consider the use of greedy plugs and b) we need to perform store queries and adjust affected snap declarations before this can be committed.

**Comment by zyga** [comment]
This interestingly panics:
```
----------------------------------------------------------------------
PANIC: managers_test.go:3341: mgrsSuite.TestUpdateManyWithAutoconnect

... Panic: change 1 unexpectedly became unready (Do) (PC=0x436E55)

/snap/go/5551/src/runtime/panic.go:967
  in gopanic
state/change.go:342
  in Change.taskStatusChanged
state/task.go:209
  in Task.SetStatus
managers_test.go:3442
  in mgrsSuite.TestUpdateManyWithAutoconnect
/snap/go/5551/src/reflect/value.go:321
  in Value.Call
/snap/go/5551/src/runtime/asm_amd64.s:1374
  in goexit
OOPS: 100 passed, 1 PANICKED
--- FAIL: TestOverlord (16.68s)
FAIL
coverage: 88.5% of statements
FAIL	github.com/snapcore/snapd/overlord	16.854s
FAIL
```

**Comment by pedronis** [comment]
@jdstrand 
> a) I'd like @pedronis to consider the use of greedy plugs

greedy plugs will connect both the system one and the snap one if the snap is installed, is that what we want?

(I don't remember the interface details) what happens if somebody has both an unconfined docker and the snap installed? what should happen?

**Comment by jdstrand** [comment]
> @jdstrand
> 
> > a) I'd like @pedronis to consider the use of greedy plugs
> 
> greedy plugs will connect both the system one and the snap one if the snap is installed, is that what we want?
> 
> (I don't remember the interface details) what happens if somebody has both an unconfined docker and the snap installed? what should happen?

That was my concern and greedy plugs are not a perfect solution for this. In this particular case for docker, it is only strictly duplicated file rules (which apparmor will dedupe) and there is no 'if classic, this policy, if not, this other policy' so it will 'work' for this. I wasn't sure if it 'working' was good enough for the PR or if we should have some sort of preference mechanism where an app snap provided slot is preferred over the system slot which is opt in on an interface by interface basis. That would work for docker (it would opt in), but I'm not sure the pattern fits; we have been trying to avoid guessing all along after all.... Perhaps there are other ideas?

**Comment by anonymouse64** [comment]
> what happens if somebody has both an unconfined docker and the snap installed? what should happen?

TLDR, only one of them will be running and usable, and this is explicitly not supported by either the docker snap or the docker deb, because they both want to listen on the same global socket, so that socket can only be provided by one or the other, but both things provide the socket at the same file location. 

Personally I'm not too worried about the implications of using the docker interface with the docker snap or the docker deb "incorrectly" (i.e. you have the docker snap installed and the docker deb, but the deb is disabled and snapd accidentally provides the docker interface through the implicit one or the inverse where the deb is enabled and the snap is disabled but snapd provides the docker interface through the docker snap) primarily because the interface provides so little in the form of just access to the socket.

We could have snapd try to do some introspection on if anything is holding the socket before deciding which one to provide, but that won't work if for example someone is _installing_ the docker snap and as such the socket isn't held by the docker snap service yet, and the deb may be disabled in which case nothing holds the socket for us to introspect. This is also a reasonable user scenario, they have the docker deb installed, and they want to use the snap, so they first disable the deb then try to install the snap and their app using the docker snap.

I think what we should do is just document very clearly in the docker interface that we should not add more policy here specifically because the docker interface as used by plugging applications may end up being connected to both the docker snap and the system snap.

**Comment by pedronis** [comment]
@jdstrand  https://github.com/snapcore/snapd/pull/7417 (which I will close for now) has a related but different situation, a case where we would like to prefer the system one.

Maybe we should have a general discussion about what to do when a system can have a system slot and a snap slot for the same interface, and the different scenarios/preferences that can apply depending on the nature of the interface/underlying resources?

**Comment by pedronis** [comment]
closing until we have a more general discussion as per previous comment

**Comment by jdstrand** [comment]
> Maybe we should have a general discussion about what to do when a system can have a system slot and a snap slot for the same interface, and the different scenarios/preferences that can apply depending on the nature of the interface/underlying resources?

@pedronis - something that occurred to me that we might consider in this conversation is allowing the user to have a voice in which is used. Eg, we try our best to have a sane default but then allow the user to use `snap set system <something>`. Other thoughts for the conversation, when the host's software and the snap are competing for a resource (eg, for the well-known docker socket location), only one can win so the admin is likely going to perform an action (eg, disable the systemd service), but I doubt we want to be in the business of detecting traditional packaging configuration (eg, the 'docker' service might be named different things depending on who packaged docker). Perhaps it makes sense to default to preferring the snap if it is installed and enabled, otherwise use implicit classic since the admin would have to take an action to fix races for resources for the snap to work at all (and maybe if there are use cases, have a 'snap set' mechanism for changing the default). None of this is a proposal and just OTOH thoughts.

**Comment by jdstrand** [comment]
We've had the conversation, so I'm re-opening (though we have to work out specific details still).

**Comment by pedronis** [comment]
> We've had the conversation, so I'm re-opening (though we have to work out specific details still).

@jdstrand I'll try to write some proposals based on that conversation in the forum when I get a bit of time (cc @AlanGriffiths )

**Comment by pedronis** [comment]
> > We've had the conversation, so I'm re-opening (though we have to work out specific details still).

> @jdstrand I'll try to write some proposals based on that conversation in the forum when I get a bit of time (cc @AlanGriffiths )

I made a RFC proposal here: https://forum.snapcraft.io/t/rfc-enforcing-system-or-app-preference-for-app-provided-slots/19358

**Comment by vilhelmprytz** [comment]
Any updates on this?
